### PR TITLE
Feat/trap focus

### PR DESCRIPTION
**Pull request by vozbrann**

### CODE REVIEW

File: src/components/trap-focus/index.tsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+// add all the elements which you want to make focusable
+const focusableElements = 'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])';
+
+export const TrapFocus = ({ children, active }: TrapFocusProps): JSX.Element => {
+  const containerRef = React.useRef(null);
+
+  const handleKeyDown = (e) => {
+    const firstElement = (containerRef?.current as HTMLElement | null)?.querySelectorAll(focusableElements)[0] || null; // get first element to be focused
```
Comment (kaliahinartem) [inline src/components/trap-focus/index.tsx:15]: don't cast values

pass generic type to useRef instead

Comment (kaliahinartem) [inline src/components/trap-focus/index.tsx:15]: || null is redundant
youo will get null anyway if there is no element

File: src/components/trap-focus/index.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    const focusableNodes = (containerRef?.current as HTMLElement | null)?.querySelectorAll(focusableElements) || [];
+    const lastElement = focusableNodes?.[focusableNodes.length - 1] || null; // get last element to be focused
```
Comment (kaliahinartem) [inline src/components/trap-focus/index.tsx:17]: || null is redundant

File: src/components/trap-focus/index.tsx [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+    const isTabPressed = e.key === 'Tab' || e.keyCode === 9;
+    if (!isTabPressed) {
+      return;
+    }
+    if (!Array.prototype.slice.call(focusableNodes).includes(document.activeElement)) {
```
Comment (kaliahinartem) [inline src/components/trap-focus/index.tsx:23]: why not Array.from(focusableNodes)?

File: src/components/trap-focus/index.tsx [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      (firstElement as HTMLElement)?.focus(); // add focus for the first focusable element
```
Comment (kaliahinartem) [inline src/components/trap-focus/index.tsx:24]: don't cast

Comment (kaliahinartem) [reply src/components/trap-focus/index.tsx:24]: the type should be inferred

File: src/components/trap-focus/index.tsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+      e.preventDefault();
+      return;
+    }
+    if (e.shiftKey) {
+      // if shift key pressed for shift + tab combination
+      if (document.activeElement === firstElement) {
+        (lastElement as HTMLElement)?.focus(); // add focus for the last focusable element
```
Comment (kaliahinartem) [inline src/components/trap-focus/index.tsx:31]: don't cast :)

File: src/components/trap-focus/index.tsx [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        e.preventDefault();
+      }
+    } else {
+      // if tab key is pressed
+      if (document.activeElement === lastElement) {
+        // if focused has reached to last focusable element then focus first focusable element after pressing tab
+        (firstElement as HTMLElement)?.focus(); // add focus for the first focusable element
```
Comment (kaliahinartem) [inline src/components/trap-focus/index.tsx:38]: aaaand don't cast :)

File: src/components/trap-focus/index.tsx [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        e.preventDefault();
+      }
+    }
+  };
+
+  React.useEffect(() => {
+    if (active) {
+      const activeElement = document.activeElement;
+      window.addEventListener('keydown', handleKeyDown);
+      (activeElement as HTMLElement)?.blur();
```
Comment (kaliahinartem) [inline src/components/trap-focus/index.tsx:48]: same

File: src/components/trap-focus/index.tsx [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+      return () => {
+        window.removeEventListener('keydown', handleKeyDown);
+        (activeElement as HTMLElement)?.focus();
```
Comment (kaliahinartem) [inline src/components/trap-focus/index.tsx:51]: same

**Review by kaliahinartem — CHANGES_REQUESTED** [review]

**Comment by kaliahinartem** [comment]
@vozbrann any updates on this PR?

**Comment by vozbrann** [comment]
> @vozbrann any updates on this PR?

@kaliahinartem Review it again, please
But maybe we should use external trap focus packages

File: src/components/popup/style.ts [hunk 9]
```diff
@@ -1,55 +1,65 @@
...
+};
 
-export const StyledPopupHeaderTitleRow = styled.div`
+export const StyledPopupHeader = styled.div<{
+  theme: PUITheme;
+}>`
   display: flex;
   justify-content: space-between;
+  align-items: center;
+  padding: 16px 24px;
```
Comment (KotAleksei) [inline src/components/popup/style.ts:18]: let's imagine if you want to customize that component, you have to change that property for each element (header, body, footer) and only once if set it to parent + check those values in figma pls(16px for mobile and 24px for desktop)

File: src/components/popup/style.ts [hunk 10]
```diff
@@ -1,55 +1,65 @@
...
+    padding: 16px;
+  }
 `;
 
-export const StyledPopupHeader = styled.div`
-  padding: 24px 24px 8px 24px;
+export const StyledPopupFooter = styled.div<{
+  theme: PUITheme;
+}>`
+  padding: 16px 24px;
```
Comment (KotAleksei) [inline src/components/popup/style.ts:28]: check distance in the design system + comment as above

File: src/components/popup/README.md [hunk 11]
```diff
@@ -17,8 +24,12 @@ const YourComponent = () => {
 
   return (
     <>
-      <PrimaryButton onClick={handlePopupShow}>Open modal</PrimaryButton>
+      <PrimaryButton onClick={handlePopupShow}>
+        Open modal
+      </PrimaryButton>
       <Popup
+        titleId="examplePopup"
```
Comment (KotAleksei) [inline src/components/popup/README.md:31]: why do you need this one ?

File: stories/popup/SmallPopup.jsx [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        <Popup.Body>
+          <div style={{marginBottom: '16px'}}>
+            Lorem ipsum dolor sit amet consectetur adipisicing elit.
+          </div>
+          <div>
+            <TextInput placeholder="Icon before" iconBefore={Icon.icons.mapPin} />
+          </div>
+        </Popup.Body>
+        <Popup.Footer>
+          <div style={{display: 'flex', justifyContent: 'flex-end', paddingBottom: '0'}}>
```
Comment (lazzystep) [inline stories/popup/SmallPopup.jsx:32]: I would move styles in `style` to class

File: stories/popup/SmallPopup.jsx [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+            <SecondaryButton aria-label="Close" onClick={handlePopupHide} style={{marginRight: '16px'}}>Cancel</SecondaryButton>
```
Comment (lazzystep) [inline stories/popup/SmallPopup.jsx:33]: I would move styles in `style` to class

File: stories/popup/SmallPopup.jsx [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+    <>
+      <div>
+        <PrimaryButton onClick={handlePopupShow}>Open small popup</PrimaryButton>
+      </div>
+      <Popup size="sm" aria-labelledby="examplePopup" show={popupOpen} onHide={handlePopupHide}>
+        <Popup.Header>
+          <Popup.Title id="examplePopup">Lorem, ipsum dolor.</Popup.Title>
+        </Popup.Header>
+        <Popup.Body>
+          <div style={{marginBottom: '16px'}}>
```
Comment (lazzystep) [inline stories/popup/SmallPopup.jsx:24]: I would move styles in `style` to class

File: stories/popup/LargePopup.jsx [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+          <div style={{marginBottom: '16px'}}>
+            <TextInput placeholder="Icon before" iconBefore={Icon.icons.mapPin} />
+          </div>
+          <div>
+            <TextArea subTitle="Subtitle" placeholder="Placeholder" />
+          </div>
+        </Popup.Body>
+        <Popup.Footer>
+          <div style={{display: 'flex', justifyContent: 'flex-end', paddingBottom: '0'}}>
+            <SecondaryButton aria-label="Close" onClick={handlePopupHide} style={{marginRight: '16px'}}>Cancel</SecondaryButton>
```
Comment (lazzystep) [inline stories/popup/LargePopup.jsx:36]: I would move styles in `style` to class

File: stories/popup/LargePopup.jsx [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    <>
+      <div>
+        <PrimaryButton onClick={handlePopupShow}>Open large popup</PrimaryButton>
+      </div>
+      <Popup size="lg" aria-labelledby="examplePopup" show={popupOpen} onHide={handlePopupHide}>
+        <Popup.Header>
+          <Popup.Title id="examplePopup">Lorem, ipsum dolor.</Popup.Title>
+        </Popup.Header>
+        <Popup.Body>
+          <div style={{marginBottom: '16px'}}>
```
Comment (lazzystep) [inline stories/popup/LargePopup.jsx:24]: I would move styles in `style` to class

File: stories/popup/LargePopup.jsx [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+            Lorem ipsum dolor sit amet consectetur adipisicing elit. Officia soluta eligendi incidunt nam neque, architecto voluptatibus! Laborum non molestiae aut.
+          </div>
+          <div style={{marginBottom: '16px'}}>
```
Comment (lazzystep) [inline stories/popup/LargePopup.jsx:27]: I would move styles in `style` to class

File: stories/popup/DefaultPopup.jsx [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+          </div>
+          <div style={{ marginBottom: '16px' }}>
+            <TextInput placeholder="Icon before" iconBefore={Icon.icons.mapPin} />
+          </div>
+          <div>
+            <TextArea subTitle="Subtitle" placeholder="Placeholder" />
+          </div>
+        </Popup.Body>
+        <Popup.Footer>
+          <div style={{ display: 'flex', justifyContent: 'flex-end', paddingBottom: '0' }}>
```
Comment (lazzystep) [inline stories/popup/DefaultPopup.jsx:36]: I would move styles in `style` to class

File: stories/popup/DefaultPopup.jsx [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+            <SecondaryButton aria-label="Close" onClick={handlePopupHide} style={{ marginRight: '16px' }}>
```
Comment (lazzystep) [inline stories/popup/DefaultPopup.jsx:37]: I would move styles in `style` to class

File: stories/popup/DefaultPopup.jsx [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+      <Popup size="md" aria-labelledby="examplePopup" show={popupOpen} onHide={handlePopupHide}>
+        <Popup.Header>
+          <Popup.Title id="examplePopup">Lorem, ipsum dolor.</Popup.Title>
+        </Popup.Header>
+        <Popup.Body>
+          <div style={{ marginBottom: '16px' }}>
+            Lorem ipsum dolor sit amet consectetur adipisicing elit. Officia soluta eligendi incidunt nam neque,
+            architecto voluptatibus! Laborum non molestiae aut.
+          </div>
+          <div style={{ marginBottom: '16px' }}>
```
Comment (lazzystep) [inline stories/popup/DefaultPopup.jsx:28]: I would move styles in `style` to class

File: stories/popup/DefaultPopup.jsx [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+    <>
+      <div>
+        <PrimaryButton onClick={handlePopupShow}>Open popup</PrimaryButton>
+      </div>
+      <Popup size="md" aria-labelledby="examplePopup" show={popupOpen} onHide={handlePopupHide}>
+        <Popup.Header>
+          <Popup.Title id="examplePopup">Lorem, ipsum dolor.</Popup.Title>
+        </Popup.Header>
+        <Popup.Body>
+          <div style={{ marginBottom: '16px' }}>
```
Comment (lazzystep) [inline stories/popup/DefaultPopup.jsx:24]: I would move styles in `style` to class

File: stories/popup/CustomSizePopup.jsx [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+        <Popup.Body>
+          <div style={{marginBottom: '16px'}}>
+            Lorem ipsum dolor sit amet consectetur adipisicing elit.
+          </div>
+          <div>
+            <TextInput placeholder="Icon before" iconBefore={Icon.icons.mapPin} />
+          </div>
+        </Popup.Body>
+        <Popup.Footer>
+          <div style={{display: 'flex', justifyContent: 'flex-end', paddingBottom: '0'}}>
```
Comment (lazzystep) [inline stories/popup/CustomSizePopup.jsx:33]: I would move styles in `style` to class

File: stories/popup/CustomSizePopup.jsx [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+            <SecondaryButton aria-label="Close" onClick={handlePopupHide} style={{marginRight: '16px'}}>Cancel</SecondaryButton>
```
Comment (lazzystep) [inline stories/popup/CustomSizePopup.jsx:34]: I would move styles in `style` to class

File: stories/popup/CustomSizePopup.jsx [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+    <>
+      <div>
+        <PrimaryButton onClick={handlePopupShow}>Open custom popup</PrimaryButton>
+      </div>
+      <Popup dialogClassName={s.customSizePopup} size="sm" aria-labelledby="examplePopup" show={popupOpen} onHide={handlePopupHide}>
+        <Popup.Header>
+          <Popup.Title id="examplePopup">Lorem, ipsum dolor.</Popup.Title>
+        </Popup.Header>
+        <Popup.Body>
+          <div style={{marginBottom: '16px'}}>
```
Comment (lazzystep) [inline stories/popup/CustomSizePopup.jsx:25]: I would move styles in `style` to class

**Review by lazzystep — CHANGES_REQUESTED** [review]
Will approve after some work on changes. Read comments

**Review by vlacher12 — APPROVED** [review]

**Comment by vlacher12** [comment]
@lazzystep

**Review by lazzystep — APPROVED** [review]

**Pull request merged**